### PR TITLE
Xmc4xxx interrupt controller

### PR DESCRIPTION
**Pull request by talih0**
This main aim of this PR is to review interrupt controller driver.
In Infineon XMC4XXX SoCs, gpio interrupts are triggered via an
Event Request Unit (ERU) module. A subset of the gpios are
connected to the ERU. The ERU monitors edge triggers and creates
a SR.
    
The interrupt controller driver configures the ERU for a target port/pin combination
for rising/falling edge events. Note that the ERU module does
not generate SR based on the gpio level. Internally the ERU
tracks the *status* of an event. The status is set on a positive
edge and unset on a negative edge (or vice-versa depending on
the configuration). The value of the status is used to implement
a level triggered interrupt; The ISR checks the status flag and
calls the callback function if the status is set.

The ERU configurations for supported port/pin combinations are
stored in a devicetree file dts/arm/infineon/xmc4xxx_x_x-intc.dtsi.
The configurations are stored in the opaque array
uint16 port_line_mapping[].

### CODE REVIEW

File: drivers/gpio/Kconfig.xmc4xxx [hunk 1]
```diff
@@ -0,0 +1,13 @@
+# XMC4XXX GPIO configuration
+#
+# Copyright (c) 2022 Schlumberger
+# SPDX-License-Identifier: Apache-2.0
+
+DT_COMPAT_INFINEON_XMC4XXX_GPIO := infineon,xmc4xxx-gpio
```
Comment (galak) [inline drivers/gpio/Kconfig.xmc4xxx:6]: Look at new Kconfig style in https://github.com/zephyrproject-rtos/zephyr/pull/48085

File: dts/bindings/gpio/infineon,xmc4xxx-gpio.yaml [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+include: [gpio-controller.yaml, base.yaml]
+
+properties:
+    reg:
+      required: true
+
+    "#gpio-cells":
+      const: 2
+
+    port-id:
```
Comment (galak) [inline dts/bindings/gpio/infineon,xmc4xxx-gpio.yaml:14]: Can't you easily "compute" `port-id` by doing simple math on the the REG_ADDR?  as such, I think you should remove `port-id` and just make a macro in the code to do the math.

**Review by galak — DISMISSED** [review]
Some Kconfig and dts updates requested.

Made the comment on gpio/Kconfig.xmc4xxx, but also applies to interrupt-controller/Kconfig.xmc4xxx.

File: dts/bindings/gpio/infineon,xmc4xxx-gpio.yaml (hunk 2 above)
Comment (talih0) [reply dts/bindings/gpio/infineon,xmc4xxx-gpio.yaml:14]: > Can't you easily "compute" `port-id` by doing simple math on the the REG_ADDR? as such, I think you should remove `port-id` and just make a macro in the code to do the math.

yes true, will update.

**Comment by cfriedt** [comment]
Can I suggest keeping this PR in draft state, since it's dependent on multiple PRs being merged first?

File: dts/bindings/gpio/infineon,xmc4xxx-gpio.yaml (hunk 2 above)
Comment (talih0) [reply dts/bindings/gpio/infineon,xmc4xxx-gpio.yaml]: I removed port-id binding in the latest version and used DT_HAS_xx_ENABLED for gpio and interrupt controller.

**Comment by cfriedt** [comment]
This PR probably shouldn't be assigned to the GPIO maintainer. However, the interrupt-controller driver area is lacking a maintainer currently. Any suggestions?

**Comment by talih0** [comment]
ping

**Comment by MaureenHelm** [comment]
Needs rebase

**Comment by talih0** [comment]
> Needs rebase

Thanks, I've rebased. I also added a fix for a special case where the gpio level was set active before the interrupt was configured.

File: dts/arm/infineon/xmc4500_F100x1024-intc.dtsi [hunk 3]
```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright (c) 2022 Schlumberger
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/* File auto-generated with python script */
```
Comment (galak) [inline dts/arm/infineon/xmc4500_F100x1024-intc.dtsi:6]: what is auto generating this file?

**Comment by galak** [comment]
> This PR probably shouldn't be assigned to the GPIO maintainer. However, the interrupt-controller driver area is lacking a maintainer currently. @galak - any suggestions?

No suggestion.

File: dts/arm/infineon/xmc4500_F100x1024-intc.dtsi (hunk 3 above)
Comment (talih0) [reply dts/arm/infineon/xmc4500_F100x1024-intc.dtsi:6]: It's my own python script. I'll remove the comment.

**Comment by talih0** [comment]
> > This PR probably shouldn't be assigned to the GPIO maintainer. However, the interrupt-controller driver area is lacking a maintainer currently. @galak - any suggestions?
> 
> No suggestion.

**Review by cfriedt — DISMISSED** [review]
GPIO changes look fine

File: drivers/gpio/gpio_xmc4xxx.c [hunk 4]
```diff
@@ -173,6 +211,9 @@ static const struct gpio_driver_api gpio_xmc4xxx_driver_api = {
 	.port_clear_bits_raw = gpio_xmc4xxx_clear_bits_raw,
 	.port_toggle_bits = gpio_xmc4xxx_toggle_bits,
 	.pin_interrupt_configure = gpio_xmc4xxx_pin_interrupt_configure,
+#if defined(CONFIG_XMC4XXX_INTC)
```
Comment (cfriedt) [inline drivers/gpio/gpio_xmc4xxx.c:214]: Here, you could use `IS_ENABLED(CONFIG_XMC4XXX_INTC) ? manage_callback : NULL` which would require fewer `#ifdef`s

**Review by cfriedt — DISMISSED** [review]
Just a general comment to optionally reduce `#ifdef`s. Otherwise the GPIO code looks fine to me

File: drivers/gpio/gpio_xmc4xxx.c (hunk 4 above)
Comment (talih0) [reply drivers/gpio/gpio_xmc4xxx.c]: thanks, I've updated the PR.

**Review by cfriedt — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]
LGTM

**Comment by talih0** [comment]
@carlescufi, @cfriedt, thank you for reviewing

**Pull request merged**